package/MDAnalysis/lib/c_gridsearch.pyx

```diff
@@ -0,0 +1,647 @@
+                sizeof(ns_neighborhood) * ncoords)
+        abort()
+
+    # Here starts the real core and the iteration over coordinates
+    for coordid in range(ncoords):
+        holder.neighborhoods[coordid] = retrieve_neighborhood(&refcoords[coordid, XX],
+                                                              neighborcoords,
+                                                              grid,
+                                                              box,
+                                                              cutoff2)
+        holder.neighborhoods[coordid].cutoff = cutoff
+
+    return holder
+
+# Python interface
+cdef class FastNS(object):
+    cdef PBCBox box
+    cdef readonly int nthreads
+    cdef readonly real[:, ::1] coords
+    cdef real[:, ::1] coords_bbox
+    cdef readonly real cutoff
+    cdef bint prepared
+    cdef ns_grid *grid
+
+
+    def __init__(self, box):
+        if box.shape != (3, 3):
+            raise ValueError("Box must be provided as triclinic_dimensions (a 3x3 numpy.ndarray of unit cell vectors")
+
+        self.box = PBCBox(box)
+
+        self.nthreads = 1
+
+        self.coords = None
+        self.coords_bbox = None
+
+        self.cutoff = -1
+
+        self.prepared = False
+
+        self.grid = <ns_grid *> malloc(sizeof(ns_grid))
+
+
+    def __dealloc__(self):
+        #destroy_nsgrid(self.grid)
+        self.grid.size = 0
+    
+    def set_coords(self, real[:, ::1] coords):
+        self.coords = coords
+
+        # Make sure atoms are inside the brick-shaped box
+        self.coords_bbox = self.box.fast_put_atoms_in_bbox(coords)
+
+        self.prepared = False
+
+
+    def set_cutoff(self, real cutoff):
+        self.cutoff = cutoff
+
+        self.prepared = False
+
+
+    def prepare(self, force=False):
+        cdef ns_int i
+        cdef bint initialization_ok
+
+        if self.prepared and not force:
+            print("NS already prepared, nothing to do!")
+
+        if self.coords is None:
+            raise ValueError("Coordinates must be set before NS preparation!")
+
+        if self.cutoff < 0:
+            raise ValueError("Cutoff must be set before NS preparation!")
+
+        with nogil:
+            initialization_ok = False
+
+            # Initializing grid
+            for i in range(DIM):
+                self.grid.ncells[i] = <ns_int> (self.box.c_pbcbox.box[i][i] / self.cutoff)
+                if self.grid.ncells[i] == 0:
+                    self.grid.ncells[i] = 1
+                self.grid.cellsize[i] = self.box.c_pbcbox.box[i][i] / self.grid.ncells[i]
+
+            self.grid.size = self.grid.ncells[XX] * self.grid.ncells[YY] * self.grid.ncells[ZZ]
+
+            # Populating grid
+            if populate_grid(self.grid, self.coords_bbox) == RET_OK:
+                initialization_ok = True
+
+
+        if initialization_ok:
+            self.prepared = True
+        else:

```

Comment:
If we reach that point, we are in a stale state. There should be some cleaning here. Though, just destroying the grid is not enough. The grid should be restored in a pre-`prepared` state.